### PR TITLE
Feature/canvas improvements

### PR DESCRIPTION
**Pull request by 95kach**
- [x] we cannot take a tab from a panel to an empty panel
- [x] when I drag a tab from one to another panel, the dragged tab should be open

https://github.com/capsulajs/capsulahub-ui/issues/8

### CODE REVIEW

File: src/components/canvas/canvas.js [hunk 1]
```diff
@@ -35,27 +35,39 @@ export default class Canvas extends React.Component {
...
+      : bus.emit('move', metadata);
+    bus.emit('dragend', {});
+  };
 
+  componentDidMount() {
     this.eventsSubscription = bus.getEventsStream(ReactDOM.findDOMNode(this)).subscribe(([event, metadata]) => {
       switch (event) {
-        case 'drop':
-          return onUpdate(createNode(this.props.layout, metadata));
+        case 'dragstart':
```
Comment (dmitriyslaym) [inline src/components/canvas/canvas.js:52]: case 'dragstart',
case 'dragover',
case 'dragend':
    return this.setState({ metadata });

File: src/components/canvas/services/bus.js [hunk 2]
```diff
@@ -22,14 +39,48 @@ export class CanvasEventBus {
...
+      map((updates) => [updates[0][0], updates.map(([_, metadata]) => metadata)])
+    );
+  }
+
+  getTabDragAndDropEventsStream() {
+    return this.subject.asObservable().pipe(filter(([event, metadata]) => ['dragstart', 'dragend'].includes(event)));
+  }
+
+  getTabReorderEventsStream() {
+    return this.subject.asObservable().pipe(filter(([event, metadata]) => ['reorder'].includes(event)));
```
Comment (dmitriyslaym) [inline src/components/canvas/services/bus.js]: all of these filter functions are very similar, I would create a common helper

File: src/components/canvas/services/index.js [hunk 3]
```diff
@@ -1 +1,5 @@
-export { CanvasEventBus } from './bus';
+import { CanvasEventBus } from './bus';
+
+const bus = new CanvasEventBus();
+
+export default bus;
```
Comment (dmitriyslaym) [inline src/components/canvas/services/index.js]: export default new CanvasEventBus();

File: src/components/canvas/tab.js [hunk 4]
```diff
@@ -26,53 +48,63 @@ const Input = styled.input`
...
       </Title>
     );
   }
+
+  render() {
+    const { isHover, isEdit } = this.state;
+    const { isActive, tab } = this.props;
+
+    return (
+      <Container isActive={isActive} onMouseEnter={() => this.onHover(true)} onMouseLeave={() => this.onHover(false)}>
```
Comment (dmitriyslaym) [inline src/components/canvas/tab.js]: It's better not to create new functions as props on each render

File: src/components/canvas/tab.js [hunk 5]
```diff
@@ -26,53 +48,63 @@ const Input = styled.input`
...
+        {this.renderContent()}
+        {!isEdit && (
+          <Close isHover={isHover} onClick={this.onRemove}>
+            ✕
```
Comment (dmitriyslaym) [inline src/components/canvas/tab.js]: using icon is better - much more flexible

**Review by dmitriyslaym — APPROVED** [review]

File: src/components/canvas/canvas.js (hunk 1 above)
Comment (95kach) [reply src/components/canvas/canvas.js:52]: #prettier

**Pull request merged**